### PR TITLE
Adding jaxtyping and small shape improvements to inference

### PR DESCRIPTION
**Pull request by mattlevine22**
- Add jaxtyping across Filter/Smoother inference.
- Canonicalize scalar observation/control series from (T,) to (T, 1).
- Preserve (T,) scalar labels for HMMs.

### CODE REVIEW

File: dynestyx/inference/filters.py [hunk 1]
```diff
@@ -275,6 +276,16 @@ def _add_log_factors(
...
+        if not isinstance(config, HMMConfigs):
+            obs_values = _ensure_trailing_event_axis(
+                obs_values,
+                plate_shapes=plate_shapes,
+            )
+            if ctrl_values is not None:
+                ctrl_values = _ensure_trailing_event_axis(
+                    ctrl_values,
+                    plate_shapes=plate_shapes,
+                )
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]: Codex points out (and I agree) that this can be problematic to do *before* plating; for plate size `M` and time series length `T`, if `M = T`, then shared scalar controls `(T,)` get promoted to plated controls `(T, 1)`, which gets interpreted as plated. We should either do this later, or raise an error in `_ensure_trailing_event_axis` for this (rather rare) edge case.

**Review by DanWaxman — CHANGES_REQUESTED** [review]

File: dynestyx/inference/integrations/cd_dynamax/continuous_filter.py [hunk 2]
```diff
@@ -136,13 +139,13 @@ def _run_linear_kf(
...
-    obs_times: jax.Array,
-    obs_values: jax.Array,
-    ctrl_times=None,
-    ctrl_values=None,
-):
+    obs_times: Real[Array, " obs_time"],
+    obs_values: Real[Array, "obs_time observation_dim"],
+    ctrl_times: Real[Array, " ctrl_time"] | None = None,
+    ctrl_values: Real[Array, "ctrl_time control_dim"] | None = None,
+) -> Any:
```
Comment (DanWaxman) [inline dynestyx/inference/integrations/cd_dynamax/continuous_filter.py:148]: I'd like to avoid `Any` as return types. This need not be blocking for now, though (I think it's a bit annoying because `cd_dynamax` types are a bit all over the place).

**Review by DanWaxman — APPROVED** [review]
I made a few tiny nitpicks and pushed. LGTM otherwise, please check my changes then feel free to merge!

File: dynestyx/inference/integrations/cd_dynamax/continuous_filter.py (hunk 2 above)
Comment (mattlevine22) [reply dynestyx/inference/integrations/cd_dynamax/continuous_filter.py:148]: agreed, but yeah let's table it. Thanks!
